Source/dom/components/Library.js

```diff
@@ -0,0 +1,141 @@
+import { WrappedObject, DefinedPropertiesKey } from '../WrappedObject'
+import { Document } from './Document'
+import { toArray, getURLFromPath } from '../utils'
+import { Types } from '../enums'
+import { Factory } from '../Factory'
+import { wrapObject } from '../wrapNativeObject'
+import { ShareableObject } from './ShareableObject'
+
+const AddStatus = {
+  0: 'ok',
+  1: 'the library has already been added',
+  2: 'the document is not in the new JSON format',
+  3: 'there was a problem reading the asset library file',
+}
+
+/**
+ * A Sketch Library.
+ */
+export class Library extends WrappedObject {
+  constructor(library = {}) {
+    if (!library.sketchObject) {
+      throw new Error('Cannot create a new Library directly')
+    }
+
+    super(library)
+  }
+
+  static getLibraries() {
+    const libraryController = AppController.sharedInstance().librariesController()
+    return toArray(libraryController.libraries()).map(
+      Library.fromNative.bind(Library)
+    )
+  }
+
+  static createLibraryFromDocument(document, path) {
```

Comment:
I think we'd be better giving this method a different name - e.g. `addUserLibraryFromDocument`. I worry that `create` implies that we are doing something to the document to turn it into a library.

Reply:
It might also be worth not passing in both `document` and `path`. If we made this `addUserLibraryFromDocumentAtPath` it would avoid the dance we're doing saving `MSDocument` / `MSDocumentData`.

Reply:
> I worry that `create` implies that we are doing something to the document to turn it into a library.

I can see where you are coming from. But at the same time, `addUserLibraryFromDocument` doesn't convey that a Library object will be returned.

A `Library` is a different object than a `Document` in the API so I don't think `createLibraryFromDocument` means that it will mutate the document in any way.

> It might also be worth not passing in both document and path.

I'm not sure I understand. and I want a user to be able to create a library from a document that is not yet saved (hence without a document existing at the path)

Reply:
> I'm not sure I understand. and I want a user to be able to create a library from a document that is not yet saved (hence without a document existing at the path)

As I read the code, at the moment you can pass in a path and `nil` as the second argument or a document and a path. But what if you pass in just the document? At the moment it looks like it would try to save with a `nil` path and then fail. If I've already saved the document before that would confuse me.

I don't think it hurts to separate this. If the method only takes a path then it's clear you need to save your document before you add it.

> at the same time, `addUserLibraryFromDocument` doesn't convey that a Library object will be returned.

Maybe it shouldn't be? You could have another method instead which returns the library at a given URL. You could then update this method so that it doesn't throw if there's already a library at that path (`addAssetLibraryAtURL:` returning `MSAssetLibraryControllerAddStatusDuplicate` - maybe the you could just return `false` in that case.) That way the method would only throw an error on more unusual cases (legacy or invalid documents at the path).

At the moment:

```
createLibraryFromDocument( doc1, path )
... followed by ...
createLibraryFromDocument( path )
... or
createLibraryFromDocument( doc2, path )
... or
createLibraryFromDocument( doc1 )
```
Would throw an error. Whereas:
```
doc1.save(path)
addUserLibraryFromDocumentAtPath(path)
lib = libraryWithPath(path)
...
doc2.save(path)
addUserLibraryFromDocumentAtPath(path)
lib = libraryWithPath(path)
```
Would leave everything in a good state regardless or the order in which they are called.

Reply:
> At the moment it looks like it would try to save with a nil path and then fail

That's fine. It's a required argument so if you don't provide it, it should fail.

> it doesn't throw if there's already a library at that path

Good point, it should throw before that, when trying to save the document if there is already one at the same path.

When I look at the use cases:
1) create a document to use as a library
2) use an existing document that is not yet a library
3) use an existing library

So:
1) `createLibraryFromDocument` is perfect for that. And you want the method to throw when there is already a document where you want to save it (while you probably don't want to throw when saving the document directly)
2) Again pretty good. You want it to throw if it's already a library
3) you just don't use `createLibraryFromDocument` and access them directly

I don't see any other use cases that would require splitting this method into 3.

Reply:
I have to admit this and the `MSDocumentData` / `MSDocument` `save` hack still doesn't sit comfortably with me.

What if we made it `libraryForDocumentAtPath` and have that method either return the existing library at that path or create and return a new one? That would leave you with 2 methods rather than 3, but I think "save this as a new document" and "use this document as a library" are 2 distinct things.